hamp_radar/iquick.py

```diff
@@ -233,6 +232,159 @@ def decode_time(ds):
+    import pandas as pd
+
+    first = read_iq(files[0])
+    last  = read_iq(files[-1])
+    tbeg  = first.time[0]
+    fbeg  = first.frm[0]
+    nfrms = last.frm[-1]-fbeg+1
+    nfft  = last.sizes['fft']
+    ngate = last.sizes['range']
+
+    tau_frame = (last.time[-1] - tbeg)/nfrms
+    tau_pulse = tau_frame/nfft
+    dtfft     = tau_pulse * xr.DataArray(np.arange(nfft), dims=("fft",))
+    len_gate  = c_light/ref_air * pulse_length / 2.
+
+    for file in files:
+        radar = read_iq(file)
+        print (f"First frame {radar.frm[0].values} of file {file}")
+        if (np.all(np.diff(radar.time)>0)) :
+            time_frame = tbeg + tau_frame*(radar.frm-fbeg)
+            ds1 = (radar[["TPow","NPw","CPw"]]
+                   .assign_coords(frame=radar.frm,frame_time=radar.time)
+                   )
+            fout  = f'{zarr_path}/HALO-{pd.to_datetime(radar.time[0].values):%Y%m%da-%H%M%S}-frms.zarr'
+            ds1.to_zarr(fout)
+
+            ds2 = (radar['FFTD'].to_dataset(dim='cocx')
+                   .assign_coords(pulse_time=lambda ds2: time_frame + dtfft)
+                   .stack(pulse_time=["frame", "fft"], create_index=False)
+                   .transpose("iq","range","pulse_time")
+                   .rename({0:"co"}).rename({1:"cx"})
+                   .assign_coords(range=("range",np.arange(ngate)*len_gate+d_gate1))
+                   .assign_coords(iq=("iq",["i","q"]))
+              )
+            ds2["range"].attrs["long_name"] = "distance from aircraft"
+            ds2["range"].attrs["units"] = "m"
+            ds2["iq"].attrs["long_name"] = "signal phase (inphase or quadrature)"
+            fout  = f'{zarr_path}/HALO-{pd.to_datetime(radar.time[0].values):%Y%m%da-%H%M%S}-cocx.zarr'
+            ds2.to_zarr(fout)
+        else:
+            sys.exit('error frames do not increment monotonically')
+    return ds_cocx,ds_frms
+
+
+def merge_iqf(cocx_files,frms_files):
+
+    """
+    Takes a list of pulse data (cocx_files) and rame data (frame_files)
+    merges them into a single data set including additional metadata, which
+    the function returns, ideally to be written to a zarr dataset.
+
+    cocx_files: list of zarr files with pulse data
+    frms_files: list of zarr files with frame data
+
+    ex.,
+    zarr_path = './radar_zarr_files'
+    files = [pds_path + x for x in sorted(os.listdir(path))]
+    untangle_iqf(files,zarr_path)
+
+    cocx_files = sorted(glob.glob(zarr_path + '*cocx.zarr'))
+    frms_files = sorted(glob.glob(zarr_path + '*frms.zarr'))
+    ds = merge_radar_zarr(cocx_files,frms_files)
+    ds.to_zarr(fout)
+    """
+    import numpy as np
+    import xarray as xr
+    ds_frms = xr.open_mfdataset(frms_files,engine='zarr')
+    dx = (ds_frms
+          .merge(ds_frms.NPw.to_dataset(dim='cocx').rename({0:"co_NPw"}).rename({1:"cx_NPw"}))
+          .merge(ds_frms.CPw.to_dataset(dim='cocx').rename({0:"co_CPw"}).rename({1:"cx_CPw"}))
+          .drop_vars(["NPw","CPw"])
+    )
```

Comment:
In case the two functions would indeed be two separate steps, this variable mangling should probably still be done in the first step. It seems unintuitive that some mangling is done in one function and some is done in the other one.